### PR TITLE
NL Performance Improvements and Coverage Improvements

### PR DESCRIPTION
**Pull request by dallasfoster**
# ALCHEMI Toolkit Pull Request

## Description

This MR primarily adds performance improvements to the neighbor list hook to cache and preallocate arrays. This operation is not quite graphable yet but could feasibly be so with some changes to the neighbor list functions in toolkit-ops. 

Test coverage is improved to around 83% in local testing.

## Type of Change

- [ x ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ x ] Performance improvement
- [ ] Documentation update
- [ x ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Testing

- [ x ] Unit tests pass locally (`make pytest`)
- [ x ] Linting passes (`make lint`)
- [ x ] New tests added for new functionality meets coverage expectations?

## Checklist

- [ x ] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [ x ] I have performed a self-review of my code
- [ x ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/dynamics/hooks/neighbor_list.py [hunk 1]
```diff
@@ -25,14 +25,68 @@
...
 (shape ``(E, 3)``, PBC only) are always up to date.  The companion
 ``Batch.edge_ptr`` property derives the per-atom CSR pointer on demand.
+
+Pre-allocation
+--------------
+The hook maintains *staging buffers* — persistent GPU tensors that are
+refreshed each step via ``Tensor.copy_()`` — to avoid per-step dynamic
+allocation inside the ``neighbor_list`` dispatcher.
+
+``neighbor_list`` selects between ``batch_naive`` (avg < 2000 atoms/system)
```
Comment (laserkelvin) [inline nvalchemi/dynamics/hooks/neighbor_list.py:35]: This might complicate things a bit, but could be good to add a link to the Toolkit Ops docs;

benchmarks: https://nvidia.github.io/nvalchemi-toolkit-ops/benchmarks/neighborlist.html
docs: https://nvidia.github.io/nvalchemi-toolkit-ops/userguide/components/neighborlist.html

File: nvalchemi/dynamics/integrators/npt.py [hunk 2]
```diff
@@ -377,7 +377,7 @@ def post_update(self, batch: Batch) -> None:
         """
         M = batch.num_graphs
         volumes = self._compute_volumes(batch)
-        cells_inv = torch.linalg.inv(batch.cell)
+        cells_inv = torch.linalg.inv_ex(batch.cell)[0].contiguous()
```
Comment (laserkelvin) [inline nvalchemi/dynamics/integrators/npt.py:380]: Why is the `contiguous` needed?

Also IMO would prefer `cells_inv, _ = torch.linalg.inv_ex(batch.cell)` as a matter of style instead of the indexing

File: nvalchemi/dynamics/integrators/nph.py [hunk 3]
```diff
@@ -220,7 +220,7 @@ def pre_update(self, batch: Batch) -> None:
         """
         volumes = self._compute_volumes(batch)
         # Compute cells_inv for velocity and position updates.
-        cells_inv = torch.linalg.inv(batch.cell)
+        cells_inv = torch.linalg.inv_ex(batch.cell)[0].contiguous()
```
Comment (laserkelvin) [inline nvalchemi/dynamics/integrators/nph.py:223]: Same comments as in NPT: why contiguous, and writing it as 

```
cells_inv, _ = torch.linalg.inv_ex(batch.cell)
cells_inv = cells_inv.contiguous()    # if it's actually needed
``` 

That said, do we need to guard against singular cells, say in a `not torch.compiler.is_compiling` setting?

File: nvalchemi/dynamics/hooks/neighbor_list.py [hunk 4]
```diff
@@ -84,101 +146,177 @@ class NeighborListHook:
...
+        # Skin check: decide per-system whether the neighbor list needs
+        # rebuilding based on atomic displacement since the last build.
+        # Uses the in-place variant to avoid per-step allocation of the
+        # rebuild_flags tensor.  Falls back to the upstream function if the
+        # in-place op is not available (nvalchemiops < 0.4 or custom op not
+        # loaded).
+        # ------------------------------------------------------------------
+        if self.skin > 0.0 and self._ref_positions is not None:
+            cell_inv = (
+                torch.linalg.inv_ex(self._buf_cell)[0].contiguous()
```
Comment (laserkelvin) [inline nvalchemi/dynamics/hooks/neighbor_list.py:256]: Would prefer not to inline just to make it easier to read

File: nvalchemi/dynamics/hooks/neighbor_list.py [hunk 5]
```diff
@@ -194,3 +332,264 @@ def _rebuild(self, batch: Batch) -> None:
...
+    # ------------------------------------------------------------------
+
+    @torch.compiler.disable
+    def _alloc_output_tensors(
+        self,
+        N: int,
+        device: torch.device,
+        pbc: torch.Tensor | None,
+    ) -> None:
+        """Allocate neighbor-matrix output tensors for atom count *N*.
```
Comment (laserkelvin) [inline nvalchemi/dynamics/hooks/neighbor_list.py:343]: I know this is a private method, but having parameter descriptions would be desirable for maintenance

File: nvalchemi/dynamics/hooks/neighbor_list.py [hunk 6]
```diff
@@ -194,3 +332,264 @@ def _rebuild(self, batch: Batch) -> None:
...
+            Staging buffer for batch_ptr (used to get max_atoms_per_system).
+        cell, pbc : torch.Tensor or None
+            Cell and PBC flag tensors at alloc time.
+        device, dtype : torch.device, torch.dtype
+            Allocation target.
+        """
+        self._buf_nl_kwargs = {}
+
+        avg_atoms = N // max(B, 1)
+        use_cell_list = avg_atoms >= 2000
```
Comment (laserkelvin) [inline nvalchemi/dynamics/hooks/neighbor_list.py:513]: Do we want to make this a user configurable thing?

Doesn't need to be in this PR, but gauging interest

File: nvalchemi/dynamics/hooks/neighbor_list.py [hunk 7]
```diff
@@ -84,101 +146,177 @@ class NeighborListHook:
...
+        is a Python mutation that creates a graph break.  Called only on the
+        first step for a given batch shape.
+        """
+        self._ref_positions = positions.detach().clone()
 
     # ------------------------------------------------------------------
     # Neighbor list construction
     # ------------------------------------------------------------------
 
     def _rebuild(self, batch: Batch) -> None:
```
Comment (laserkelvin) [inline nvalchemi/dynamics/hooks/neighbor_list.py:213]: Redundant...?

File: nvalchemi/dynamics/hooks/neighbor_list.py [hunk 8]
```diff
@@ -84,101 +146,177 @@ class NeighborListHook:
...
+        self._buf_pbc: torch.Tensor | None = None  # PBC only
+
+        # Algorithm-specific pre-allocated kwargs forwarded to neighbor_list.
+        self._buf_nl_kwargs: dict[str, torch.Tensor] = {}
 
     # ------------------------------------------------------------------
     # Main hook entry point
     # ------------------------------------------------------------------
-
+    @torch.compile(fullgraph=False, mode="max-autotune-no-cudagraphs")
```
Comment (laserkelvin) [inline nvalchemi/dynamics/hooks/neighbor_list.py:183]: I wonder if this `torch.compile` makes a big difference?

File: nvalchemi/dynamics/hooks/neighbor_list.py [hunk 9]
```diff
@@ -194,3 +332,264 @@ def _rebuild(self, batch: Batch) -> None:
...
+            else None,
+            fill_value=batch.num_nodes,
+        )
+        edge_index = neighbor_list_coo[0].T.contiguous()  # (E, 2) int32
+        if len(neighbor_list_coo) > 2:
+            unit_shifts = neighbor_list_coo[2].to(torch.int32)  # (E, 3) int32
+        else:
+            unit_shifts = None
+
+        from nvalchemi.data.level_storage import SegmentedLevelStorage
```
Comment (laserkelvin) [inline nvalchemi/dynamics/hooks/neighbor_list.py:445]: Move to top

File: test/data/test_level_storage.py [hunk 10]
```diff
@@ -1051,3 +1051,317 @@ def test_default_attribute_map_has_expected_groups(self):
...
+        schema = LevelSchema()
+        with pytest.raises(KeyError, match="not found"):
+            schema.is_segmented_attr("nonexistent_attr_xyz")
+
+
+# -----------------------------------------------------------------------------
+# UniformLevelStorage.extend_for_appended_graphs
+# -----------------------------------------------------------------------------
+class TestUniformLevelStorageExtend:
+    """Tests for UniformLevelStorage.extend_for_appended_graphs (lines 1036–1054)."""
```
Comment (laserkelvin) [inline test/data/test_level_storage.py]: I would remove the line counts, and for one or two of these consider adding GPU tests as well

File: test/data/test_reader_base.py
Comment (laserkelvin) [inline test/data/test_reader_base.py:1]: Is this just because of coverage?

File: test/data/test_zarr_datapipe.py [hunk 11]
```diff
@@ -1982,3 +1982,194 @@ def test_append_dispatches_to_memory_store_set(self, num_samples: int) -> None:
...
+    # ------------------------------------------------------------------
+    # Construction edge-cases
+    # ------------------------------------------------------------------
+
+    def test_invalid_reader_raises_type_error(self):
+        """Passing an object that doesn't implement ReaderProtocol raises TypeError."""
+        with pytest.raises(TypeError, match="Reader interface"):
+            Dataset(object())  # type: ignore[arg-type]
+
+    def test_device_string_is_converted_to_torch_device(self):
```
Comment (laserkelvin) [inline test/data/test_zarr_datapipe.py]: Should parametrize for device

File: test/data/test_zarr_datapipe.py [hunk 12]
```diff
@@ -1982,3 +1982,194 @@ def test_append_dispatches_to_memory_store_set(self, num_samples: int) -> None:
...
+    # ------------------------------------------------------------------
+
+    def test_getitem_returns_atomic_data_and_metadata(self):
+        reader = _SimpleReader()
+        ds = Dataset(reader, device="cpu")
+        data, meta = ds[0]
+        assert isinstance(data, AtomicData)
+        assert "src_index" in meta
+
+    def test_getitem_transfers_to_target_device(self):
```
Comment (laserkelvin) [inline test/data/test_zarr_datapipe.py]: Should definitely parametrize for device

File: test/models/test_mace.py [hunk 13]
```diff
@@ -806,15 +806,15 @@ def test_compile_inference(self):
...
-        except Exception as e:
-            if "NotImplementedError" in str(e) or "SEQUENCE_LENGTH" in str(e):
-                pytest.skip(
-                    "torch.compile + MACE failed (e3nn Irreps guard issue); "
-                    "needs MACE patch from mace-org/mace@6a32999"
-                )
-            raise
+        # try:
+        out = w.forward(batch)
+        # except Exception as e:
```
Comment (laserkelvin) [inline test/models/test_mace.py]: This got commented out?

File: nvalchemi/dynamics/integrators/npt.py (hunk 2 above)
Comment (dallasfoster) [reply nvalchemi/dynamics/integrators/npt.py:380]: I believe the strides are changed in this operation because warp will error in wp.from_torch unless the inverse cell is made contiguous.

File: nvalchemi/dynamics/integrators/nph.py (hunk 3 above)
Comment (dallasfoster) [reply nvalchemi/dynamics/integrators/nph.py:223]: I will consider the singular cell feature.

File: nvalchemi/dynamics/hooks/neighbor_list.py (hunk 6 above)
Comment (dallasfoster) [reply nvalchemi/dynamics/hooks/neighbor_list.py:513]: Good point, will see. I think immediately I was just trying to reflect the toolkit-ops usage.

File: nvalchemi/dynamics/hooks/neighbor_list.py (hunk 8 above)
Comment (dallasfoster) [reply nvalchemi/dynamics/hooks/neighbor_list.py:183]: Turns out, marginally.

File: test/data/test_reader_base.py
Comment (dallasfoster) [reply test/data/test_reader_base.py:1]: Yes.

File: test/models/test_mace.py (hunk 13 above)
Comment (dallasfoster) [reply test/models/test_mace.py]: :eek:
